src/components/organisms/ConceptPicker.tsx

```diff
@@ -0,0 +1,167 @@
+
+import { QUERY_PARAMS } from "@/constants/queryParams";
+
+import { TConcept } from "@/types";
+
+type TProps = {
+  concepts: TConcept[];
+  startingSort?: TSort;
+};
+
+type TSort = "A-Z" | "Grouped";
+
+const SORT_OPTIONS: TSort[] = ["A-Z", "Grouped"];
+
+const isSelected = (queryValue: string | string[] | undefined, option: string) => {
+  if (!queryValue) {
+    return false;
+  }
+  if (typeof queryValue === "string") {
+    return queryValue === option;
+  } else {
+    return queryValue.includes(option);
+  }
+};
+
+const conceptsSorter = (a: TConcept, b: TConcept, sort: TSort) => {
+  if (sort === "A-Z") {
+    return a.preferred_label.localeCompare(b.preferred_label);
+  } else if (sort === "Grouped") {
+    return a.recursive_subconcept_of[0].localeCompare(b.recursive_subconcept_of[0]);
+  }
+  return 0;
+};
+
+const filterConcepts = (concepts: TConcept[], search: string) => {
+  return concepts.filter(
+    (concept) =>
+      concept.preferred_label.toLowerCase().includes(search.toLowerCase()) ||
+      concept.alternative_labels.some((label) => label.toLowerCase().includes(search.toLowerCase()))
+  );
+};
+
+const onConceptChange = (router: NextRouter, concept: TConcept) => {
+  const query = { ...router.query };
+  let selectedConcepts: string | string[] = query[QUERY_PARAMS.concept_name] || [];
+
+  if (Array.isArray(selectedConcepts)) {
+    if (selectedConcepts.includes(concept.preferred_label)) {
+      selectedConcepts = selectedConcepts.filter((c) => c !== concept.preferred_label);
+    } else {
+      selectedConcepts = [...selectedConcepts, concept.preferred_label];
+    }
+  } else {
+    if (selectedConcepts === concept.preferred_label) {
+      selectedConcepts = [];
+    } else {
+      selectedConcepts = [selectedConcepts, concept.preferred_label];
+    }
+  }
+
+  query[QUERY_PARAMS.concept_name] = selectedConcepts;
+
+  router.push({ query: query }, undefined, { shallow: true });
+};
+
+const getSelectedConcepts = (concepts: TConcept[], router: NextRouter) => {
+  return concepts.reduce((acc, concept) => {
+    const queryValue = router.query[QUERY_PARAMS.concept_name];
+    if (!queryValue) {
+      return acc;
+    }
+    if (typeof queryValue === "string") {
+      return queryValue === concept.preferred_label ? [...acc, concept] : acc;
+    }
+    return queryValue.includes(concept.preferred_label) ? [...acc, concept] : acc;
+  }, [] as TConcept[]);
+};
+
+export const ConceptPicker = ({ concepts, startingSort = "A-Z" }: TProps) => {
+  const router = useRouter();
+  const ref = useRef(null);
+  const [search, setSearch] = useState("");
+  const [sort, setSort] = useState<TSort>(startingSort);
+  const [rootConcepts, setRootConcepts] = useState<TConcept[]>([]);
+  const [conceptsGrouped, setConceptsGrouped] = useState<{ [rootConceptId: string]: TConcept[] }>({});
+  const [filteredConcepts, setFilteredConcepts] = useState<TConcept[]>([]);
+
+  useEffect(() => {
```

Comment:
Praise: Great use of `useEffect`, and there's lots of good function names on show here.